firmware/components/buses/buses.c

```diff
@@ -0,0 +1,318 @@
+        vspiBusConfiguration.quadwp_io_num   = CONFIG_PIN_NUM_HSPI_WP;
+        vspiBusConfiguration.quadhd_io_num   = CONFIG_PIN_NUM_HSPI_HD;
+        vspiBusConfiguration.max_transfer_sz = CONFIG_BUS_HSPI_MAX_TRANSFERSIZE;
+        res = spi_bus_initialize(HSPI_HOST, &vspiBusConfiguration, CONFIG_BUS_HSPI_DMA_CHANNEL);
+        if (res != ESP_OK) return res;
+    #endif
+
+    #ifdef CONFIG_BUS_I2C0_ENABLE
+        i2c_config_t i2c0BusConfiguration = {0};
+        i2c0BusConfiguration.mode             = I2C_MODE_MASTER;
+        i2c0BusConfiguration.sda_io_num       = CONFIG_PIN_NUM_I2C0_DATA;
+        i2c0BusConfiguration.scl_io_num       = CONFIG_PIN_NUM_I2C0_CLK;
+        i2c0BusConfiguration.master.clk_speed = CONFIG_I2C0_MASTER_FREQ_HZ,
+        #ifdef CONFIG_PIN_PULL_I2C0_DATA
+            i2c0BusConfiguration.sda_pullup_en = GPIO_PULLUP_ENABLE,
+        #else
+            i2c0BusConfiguration.sda_pullup_en = GPIO_PULLUP_DISABLE,
+        #endif
+        #ifdef CONFIG_PIN_PULL_I2C0_CLK
+            i2c0BusConfiguration.scl_pullup_en = GPIO_PULLUP_ENABLE,
+        #else
+            i2c0BusConfiguration.scl_pullup_en = GPIO_PULLUP_DISABLE,
+        #endif
+
+        res = i2c_param_config(I2C_NUM_0, &i2c0BusConfiguration);
+        if (res != ESP_OK) return res;
+        res = i2c_driver_install(I2C_NUM_0, i2c0BusConfiguration.mode, I2C_MASTER_RX_BUF_DISABLE, I2C_MASTER_TX_BUF_DISABLE, 0);
+        if (res != ESP_OK) return res;
+        
+        i2c0_mux = xSemaphoreCreateMutex();
+        if (i2c0_mux == NULL) return ESP_ERR_NO_MEM;
+    #endif
+
+    #ifdef CONFIG_BUS_I2C1_ENABLE
+        i2c_config_t i2c1BusConfiguration = {0};
+        i2c1BusConfiguration.mode             = I2C_MODE_MASTER;
+        i2c1BusConfiguration.sda_io_num       = CONFIG_PIN_NUM_I2C1_DATA;
+        i2c1BusConfiguration.scl_io_num       = CONFIG_PIN_NUM_I2C1_CLK;
+        i2c1BusConfiguration.master.clk_speed = CONFIG_I2C1_MASTER_FREQ_HZ,
+        #ifdef CONFIG_PIN_PULL_I2C1_DATA
+            i2c1BusConfiguration.sda_pullup_en = GPIO_PULLUP_ENABLE,
+        #else
+            i2c1BusConfiguration.sda_pullup_en = GPIO_PULLUP_DISABLE,
+        #endif
+        #ifdef CONFIG_PIN_PULL_I2C1_CLK
+            i2c1BusConfiguration.scl_pullup_en = GPIO_PULLUP_ENABLE,
+        #else
+            i2cqBusConfiguration.scl_pullup_en = GPIO_PULLUP_DISABLE,
+        #endif
+
+        res = i2c_param_config(I2C_NUM_1, &i2c1BusConfiguration);
+        if (res != ESP_OK) return res;
+        res = i2c_driver_install(I2C_NUM_1, i2c1BusConfiguration.mode, I2C_MASTER_RX_BUF_DISABLE, I2C_MASTER_TX_BUF_DISABLE, 0);
+        if (res != ESP_OK) return res;
+        
+        i2c1_mux = xSemaphoreCreateMutex();
+        if (i2c1_mux == NULL) return ESP_ERR_NO_MEM;
+    #endif
+
+    return ESP_OK;
+}
+
+/* I2C helper functions */
+
+esp_err_t driver_i2c_read_bytes(int bus, uint8_t addr, uint8_t *value, size_t value_len) {
+    i2c_cmd_handle_t cmd = i2c_cmd_link_create();
```

Comment:
This "cmd" structure is a description of the I2C commands to be executed, it doesn't actually use the I2C port yet here.